dialogue-core/src/main/java/com/palantir/dialogue/core/StickyEndpointChannels2.java

```diff
@@ -0,0 +1,302 @@
+
+    private final EndpointChannelFactory delegate;
+
+    private StickyEndpointChannels2(EndpointChannelFactory endpointChannelFactory) {
+        this.delegate = Preconditions.checkNotNull(endpointChannelFactory, "endpointChannelFactory");
+    }
+
+    @Override
+    public Supplier<Channel> get() {
+        return new StickySessionSupplier(delegate);
+    }
+
+    @Override
+    public String toString() {
+        return "StickyEndpointChannels2{}";
+    }
+
+    private static final class StickySessionSupplier implements Supplier<Channel> {
+
+        private final EndpointChannelFactory channelFactory;
+        private final StickyRouter router = new DefaultStickyRouter();
+
+        private StickySessionSupplier(EndpointChannelFactory channelFactory) {
+            this.channelFactory = channelFactory;
+        }
+
+        @Override
+        public Channel get() {
+            return new Sticky(channelFactory, router);
+        }
+    }
+
+    private static final class Sticky implements EndpointChannelFactory, Channel {
+
+        private final EndpointChannelFactory channelFactory;
+        private final StickyRouter router;
+        private final RoutingKey routingKey;
+
+        private Sticky(EndpointChannelFactory channelFactory, StickyRouter router) {
+            this.router = router;
+            this.channelFactory = channelFactory;
+            this.routingKey = RoutingKey.create();
+        }
+
+        @Override
+        public EndpointChannel endpoint(Endpoint endpoint) {
+            return new StickyEndpointChannel(router, channelFactory.endpoint(endpoint), routingKey);
+        }
+
+        /**
+         * .
+         * @deprecated prefer {@link #endpoint}, as it allows binding work upfront
+         */
+        @Deprecated
+        @Override
+        public ListenableFuture<Response> execute(Endpoint endpoint, Request request) {
+            return endpoint(endpoint).execute(request);
+        }
+
+        @Override
+        public String toString() {
+            return "Sticky{" + channelFactory + '}';
+        }
+    }
+
+    public static StickyEndpointChannels2 create(EndpointChannelFactory endpointChannelFactory) {
+        return new StickyEndpointChannels2(endpointChannelFactory);
+    }
+
+    interface StickyRouter {
+        ListenableFuture<Response> execute(Request request, EndpointChannel endpointChannel);
+    }
+
+    // This single-in-flight-call has been implemented a few times, this impl is likely a bit buggy,
+    // but shows the approach here.
+    // Initial implementation just blocked all calls until we have hostId, but that means simulations were not
+    // working.
+    private static final class DefaultStickyRouter implements StickyRouter {
+
+        private static final ThreadLocal<Integer> stackDepth = ThreadLocal.withInitial(() -> 0);
+        private static final ThreadLocal<Integer> numDeferred = ThreadLocal.withInitial(() -> 0);
```

Comment:
Why are we using threadlocals?

Reply:
Discussed offline: I just tried to implement this in a way to allow me to make progress, it's rather janky. What this is supposed to do is allow you to execute first request, and queue up other requests behind that one completing successfully, so they can peel out the host and route next requests to the same host. @carterkozak suggested using future callbacks to form a natural queue.

Reply:
The retrying channel is a great resource/template for this sort of thing, the trasnformAsync function works wonders. The retrying channel effectively queues retries after the initial request, conditionally upon remaining retries and the result of the request (success/failure).